dallinger/heroku/clock.py

```diff
@@ -37,11 +38,32 @@ def run_check(participants, config, reference_time):
 def check_db_for_missing_notifications():
     """Check the database for missing notifications."""
     config = dallinger.config.get_config()
-    participants = Participant.query.filter_by(status="working").all()
     reference_time = datetime.now()
-
-    run_check(participants, config, reference_time)
-    db.session.commit()
+    with db.sessions_scope() as session:
+        participants = session.query(Participant).filter_by(status="working").all()
+        run_check(participants, config, reference_time)
+
+
+@scheduler.scheduled_job("interval", seconds=10)
+def warn_on_idle_transactions():
+    """Warn if any DB sessions are idle in transaction for > 1 second."""
+    with db.engine.connect() as conn:
+        result = conn.execute(
+            text(
+                """
+            SELECT pid, state, xact_start, query, now() - xact_start AS idle_time
+            FROM pg_stat_activity
+            WHERE state = 'idle in transaction'
+              AND xact_start IS NOT NULL
+              AND now() - xact_start > interval '1 second'
+        """
+            )
+        )
+        for row in result:
+            print(
+                f"Session idle in transaction! pid={row.pid}, "
+                f"idle_time={row.idle_time}, last_query={row.query}"
+            )
```

Comment:
@jessesnyder On line 87 below, could we just change:
```
task = getattr(experiment_class, meth_name, None)
```
To:
```
task = db.scoped_session_decorator(getattr(experiment_class, meth_name, None))
```
To ensure that all experiment registered scheduled tasks get session management?

Reply:
Yes, that looks really promising and gets us out of documenting a possible foot-gun for experiment authors!

Reply:
Added